TPCClusterFinder: Add gain calibration.

fweig opened the pull request:
This PR adds gain correction of the TPC pads in the cluster finder.

@davidrohr I saw about 8-9% performance overhead in the decodeZS kernel from these changes. Using chars instead of short to pack the calibration values seems to have only a negligible effect on runtime.

fweig committed:
TPCClusterFinder: Add gain calibration.

davidrohr commented on Detectors/TPC/workflow/src/CATrackerSpec.cxx:
```diff
@@ -195,6 +199,17 @@ DataProcessorSpec getCATrackerSpec(ca::Config const& specconfig, std::vector<int
 
       config.configCalib.dEdxSplines = processAttributes->dEdxSplines.get();
 
+      auto& theMapper = o2::tpc::Mapper::instance();
+      std::string gainCorrectionFile{"GainMap_FromProductionData_round2.root"}; // TODO: set from flag
+      if (boost::filesystem::exists(gainCorrectionFile)) {
+        LOG(INFO) << "Loading tpc gain correction from file " << gainCorrectionFile;
+        const auto* gainMap = o2::tpc::utils::readCalPads(gainCorrectionFile, "GainMap")[0];
+        processAttributes->tpcCalibration.reset(new TPCCFCalibration{theMapper, *gainMap});
+      } else {
+        processAttributes->tpcCalibration.reset(new TPCCFCalibration{theMapper});
+      }
+      config.configCalib.tpcCalibration = processAttributes->tpcCalibration.get();
+
```
The file should be set in `GPUSettingsO2` part in `GPUSettingsList.h` avaiable as `confParam` in `CATrackerSpec.cxx`, and it should be optional, as is already the case.

davidrohr commented on GPU/Common/GPUCommonMath.h:
```diff
@@ -47,6 +47,11 @@ class GPUCommonMath
   template <class T, class S, class R>
   GPUd() static T MaxWithRef(T x, T y, T z, T w, S refX, S refY, S refZ, S refW, R& r);
+  template <class T>
+  GPUd() static T Clamp(const T v, const T lo, const T hi)
```
Should be GPUdi() for clarity when the implementation is inline.

davidrohr commented on Detectors/TPC/reconstruction/test/testGPUCATracking.cxx:
```diff
@@ -21,12 +21,14 @@
 #include "DataFormatsTPC/TrackTPC.h"
 #include "DataFormatsTPC/ClusterNative.h"
 #include "DataFormatsTPC/ClusterNativeHelper.h"
+#include "TPCBase/Mapper.h"
```
Just a note: using the mapper is fine in principle, but a lot of the mapper functionality is available in `SliceTracker/GPUTPCGeometry.h`, which is part of the `GPUParam`. I would take as much as possible from there.

davidrohr reviewed (changes requested):
Looks ok in general, just some minor comments below + could you add also the respective code to `DumpSettings` and `ReadSettings` in `GPUChainTrackingIO.cxx`.

fweig commented:
Pushed an update that addresses these points. Added a flag `gainCalibFile` to `GPUSettingsO2`. The `TPCCFCalibration` class now uses `GPUTPCGeometry` instead of the tpc mapper.

Edit: `GPUChainTrackingIO` also updated.

davidrohr reviewed (approved):
This looks all ok now.
- Results are identical to before when applying a trivial gain correction and still reasonable with gain correction.
- I see a performance impact of ~25-30% during the TPC ZS decoding which translates to a ~1% total performance impact for the GPU processing, which seems reasonable for the gain correction.

@wiechula @matthias-kleiner : You can use the feature by adding to the tpc-reco-workflow command line: `--configKeyValues "GPU_global.gainCalibFile=[filename]"`, while the file should be in the format as the example you sent me.
